src/tools/info/proto_info.c

```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
     printf("#\n");
 }
 
-ucs_status_t print_ucp_info(int print_opts,
-                            ucs_config_print_flags_t print_flags,
-                            uint64_t ctx_features,
-                            const ucp_ep_params_t *base_ep_params,
-                            size_t estimated_num_eps, size_t estimated_num_ppn,
-                            unsigned dev_type_bitmap, const char *mem_size)
+static void listener_accept_callback(ucp_ep_h ep, void *arg)
 {
-    ucp_config_t *config;
+    *(ucp_ep_h*)arg = ep;
+}
+
+static void
+set_saddr(const char *addr_str, uint16_t port, struct sockaddr_in *saddr)
+{
+    memset(saddr, 0, sizeof(*saddr));
+    saddr->sin_family      = AF_INET;
+    saddr->sin_addr.s_addr = inet_addr(addr_str);
+    saddr->sin_port        = htons(port);
+}
+
+static ucs_status_t
+wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
+{
+    ucs_status_t status;
+
+    if (status_ptr == NULL) {
+        status = UCS_OK;
+    } else if (UCS_PTR_IS_PTR(status_ptr)) {
+        do {
+            ucp_worker_progress(worker);
+            status = ucp_request_test(status_ptr, NULL);
+        } while (status == UCS_INPROGRESS);
+        ucp_request_release(status_ptr);
+    } else {
+        status = UCS_PTR_STATUS(status_ptr);
+    }
+
+    return status;
+}
+
+static ucs_status_t
+ep_close(ucp_worker_h worker, ucp_ep_h ep, ucp_ep_close_flags_t flags,
+         const char *ep_type)
+{
+    ucp_request_param_t request_param;
+    ucs_status_ptr_t status_ptr;
+    ucs_status_t status;
+
+    request_param.op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS;
+    request_param.flags        = flags;
+
+    status_ptr = ucp_ep_close_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to destroy %s UCP endpoint>\n", ep_type);
```

Comment:
destroy must be completed for any status excepting "in progress", return code of ep_close indicates status ro the peer, I think better to ignore it.

Reply:
peer is on the same process, so I'd expect that it shouldn't fail

Reply:
anyway better to follow the doc, IMO this error message is wrong

Reply:
https://github.com/openucx/ucx/blob/b7ddc45fccd955b604f28abd948c3f8b7e03c4c7/examples/ucp_client_server.c#L471
the code is almost the same that we have in `ucp_client_server` example

Reply:
then I think we have to fix there too :)

Reply:
fixed